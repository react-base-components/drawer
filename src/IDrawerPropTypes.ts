// Type definitions for rb-drawer 1.9
// Project: https://github.com/react-base-components/drawer
// Definitions by: jljsj33 <https://github.com/jljsj33>
// Definitions: https://github.com/react-base-components/drawer
import * as React from 'react';


export type IPlacement = 'left' | 'top' | 'right' | 'bottom';

type ILevelMove = number | [number, number];

type IStringOrHtmlElement = string | HTMLElement;

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

interface IProps extends Omit<React.HTMLAttributes<any>, 'onChange'> {
  prefixCls?: string;
  width?: string | number;
  height?: string | number;
  open?: boolean;
  defaultOpen?: boolean;
  handler?: React.ReactElement | null | false;
  placement?: IPlacement;
  level?: null | string | string[];
  levelMove?: ILevelMove | ((e: { target: HTMLElement, open: boolean }) => ILevelMove);
  duration?: string;
  ease?: string;
  showMask?: boolean;
  maskClosable?: boolean;
  maskStyle?: React.CSSProperties;
  onChange?: ((open?: boolean) => void);
  afterVisibleChange?: ((open: boolean) => void);
  onHandleClick?: ((e: React.MouseEvent | React.KeyboardEvent) => void);
  onClose?: ((e: React.MouseEvent | React.KeyboardEvent) => void);
  keyboard?: boolean;
}

export interface IDrawerProps extends IProps {
  wrapperClassName?: string;
  forceRender?: boolean;
  getContainer?: IStringOrHtmlElement | (() => IStringOrHtmlElement) | null | false;
}

export interface IDrawerChildProps extends IProps {
  getContainer?: () => HTMLElement;
  getOpenCount?: () => number;
  switchScrollingEffect?: (close?: boolean) => void;
}
