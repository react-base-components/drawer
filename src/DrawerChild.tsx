import classnames from 'classnames';
import getScrollBarSize from 'rc-util/lib/getScrollBarSize';
import KeyCode from 'rc-util/lib/KeyCode';
import * as React from 'react';
import { polyfill } from 'react-lifecycles-compat';

import { IDrawerChildProps } from './IDrawerPropTypes';

import {
  addEventListener,
  dataToArray,
  getTouchParentScroll,
  isNumeric,
  removeEventListener,
  transformArguments,
  transitionEnd,
  transitionStr,
  windowIsUndefined,
} from './utils';

const currentDrawer: {
  [key: string]: boolean;
} = {};

interface IState {
  _self: DrawerChild;
  prevProps?: IDrawerChildProps;
}

class DrawerChild extends React.Component<IDrawerChildProps, IState> {
  static defaultProps = {
    switchScrollingEffect: () => {},
  };

  public static getDerivedStateFromProps(
    props: IDrawerChildProps,
    { prevProps, _self }: { prevProps: IDrawerChildProps; _self: DrawerChild },
  ) {
    const nextState = {
      prevProps: props,
    };
    if (prevProps !== undefined) {
      const { placement, level } = props;
      if (placement !== prevProps.placement) {
        // test 的 bug, 有动画过场，删除 dom
        _self.contentDom = null;
      }
      if (level !== prevProps.level) {
        _self.getLevelDom(props);
      }
    }
    return nextState;
  }

  private levelDom: HTMLElement[];

  private dom: HTMLElement;

  private contentWrapper: HTMLElement;

  private contentDom: HTMLElement | null;

  private maskDom: HTMLElement;

  private handlerDom: HTMLElement;

  private drawerId: string;

  private timeout: any;

  private passive: { passive: boolean } | boolean;

  private startPos: {
    x: number;
    y: number;
  };

  constructor(props: IDrawerChildProps) {
    super(props);
    this.state = {
      _self: this,
    };
  }

  public componentDidMount() {
    if (!windowIsUndefined) {
      let passiveSupported = false;
      try {
        window.addEventListener(
          'test',
          null,
          Object.defineProperty({}, 'passive', {
            get: () => {
              passiveSupported = true;
              return null;
            },
          }),
        );
      } catch (err) {}
      this.passive = passiveSupported ? { passive: false } : false;
    }
    const { open } = this.props;
    this.drawerId = `drawer_id_${Number(
      (Date.now() + Math.random())
        .toString()
        .replace('.', Math.round(Math.random() * 9).toString()),
    ).toString(16)}`;
    this.getLevelDom(this.props);
    if (open) {
      currentDrawer[this.drawerId] = open;
      // level;
      this.openLevelTransition();
      this.forceUpdate(() => {
        this.domFocus();
      });
    }
  }

  public componentDidUpdate(prevProps: IDrawerChildProps) {
    const { open } = this.props;
    if (open !== prevProps.open) {
      if (open) {
        this.domFocus();
      }
      currentDrawer[this.drawerId] = !!open;
      this.openLevelTransition();
    }
  }

  public componentWillUnmount() {
    const { getOpenCount, open, switchScrollingEffect } = this.props;
    const openCount = typeof getOpenCount === 'function' && getOpenCount();
    delete currentDrawer[this.drawerId];
    if (open) {
      this.setLevelTransform(false);
      document.body.style.touchAction = '';
    }
    if (!openCount) {
      document.body.style.overflow = '';
      switchScrollingEffect(true);
    }
  }

  private domFocus = () => {
    if (this.dom) {
      this.dom.focus();
    }
  };

  private removeStartHandler = (e: React.TouchEvent | TouchEvent) => {
    if (e.touches.length > 1) {
      return;
    }
    this.startPos = {
      x: e.touches[0].clientX,
      y: e.touches[0].clientY,
    };
  };

  private removeMoveHandler = (e: React.TouchEvent | TouchEvent) => {
    if (e.changedTouches.length > 1) {
      return;
    }
    const currentTarget = e.currentTarget as HTMLElement;
    const differX = e.changedTouches[0].clientX - this.startPos.x;
    const differY = e.changedTouches[0].clientY - this.startPos.y;
    if (
      currentTarget === this.maskDom ||
      currentTarget === this.handlerDom ||
      (currentTarget === this.contentDom &&
        getTouchParentScroll(
          currentTarget,
          e.target as HTMLElement,
          differX,
          differY,
        ))
    ) {
      e.preventDefault();
    }
  };

  private transitionEnd = (e: TransitionEvent) => {
    const dom: HTMLElement = e.target as HTMLElement;
    removeEventListener(dom, transitionEnd, this.transitionEnd);
    dom.style.transition = '';
  };

  private onKeyDown = (e: React.KeyboardEvent) => {
    if (e.keyCode === KeyCode.ESC) {
      const { onClose } = this.props;
      e.stopPropagation();
      if (onClose) {
        onClose(e as any);
      }
    }
  };

  private onWrapperTransitionEnd = (e: React.TransitionEvent) => {
    const { open, afterVisibleChange } = this.props;
    if (
      e.target === this.contentWrapper &&
      e.propertyName.match(/transform$/)
    ) {
      this.dom.style.transition = '';
      if (!open && this.getCurrentDrawerSome()) {
        document.body.style.overflowX = '';
        if (this.maskDom) {
          this.maskDom.style.left = '';
          this.maskDom.style.width = '';
        }
      }
      if (afterVisibleChange) {
        afterVisibleChange(!!open);
      }
    }
  };

  private openLevelTransition = () => {
    const { open, width, height } = this.props;
    const {
      isHorizontal,
      placementName,
    } = this.getHorizontalBoolAndPlacementName();
    const contentValue = this.contentDom
      ? this.contentDom.getBoundingClientRect()[
          isHorizontal ? 'width' : 'height'
        ]
      : 0;
    const value = (isHorizontal ? width : height) || contentValue;
    this.setLevelAndScrolling(open, placementName, value);
  };

  private setLevelTransform = (
    open?: boolean,
    placementName?: string,
    value?: string | number,
    right?: number,
  ) => {
    const { placement, levelMove, duration, ease, showMask } = this.props;
    // router
    this.levelDom.forEach(dom => {
      dom.style.transition = `transform ${duration} ${ease}`;
      addEventListener(dom, transitionEnd, this.transitionEnd);
      let levelValue = open ? value : 0;
      if (levelMove) {
        const $levelMove = transformArguments(levelMove, { target: dom, open });
        levelValue = open ? $levelMove[0] : $levelMove[1] || 0;
      }
      const $value =
        typeof levelValue === 'number' ? `${levelValue}px` : levelValue;
      let placementPos =
        placement === 'left' || placement === 'top' ? $value : `-${$value}`;
      placementPos =
        showMask && placement === 'right' && right
          ? `calc(${placementPos} + ${right}px)`
          : placementPos;
      dom.style.transform = levelValue
        ? `${placementName}(${placementPos})`
        : '';
    });
  };

  private setLevelAndScrolling = (
    open?: boolean,
    placementName?: string,
    value?: string | number,
  ) => {
    const { onChange } = this.props;
    if (!windowIsUndefined) {
      const right =
        document.body.scrollHeight >
          (window.innerHeight || document.documentElement.clientHeight) &&
        window.innerWidth > document.body.offsetWidth
          ? getScrollBarSize(true)
          : 0;
      this.setLevelTransform(open, placementName, value, right);
      this.toggleScrollingToDrawerAndBody(right);
    }
    if (onChange) {
      onChange(open);
    }
  };

  private toggleScrollingToDrawerAndBody = (right: number) => {
    const { getOpenCount, getContainer, showMask, open } = this.props;
    const container = getContainer && getContainer();
    const openCount = getOpenCount && getOpenCount();
    // body 
    if (container && container.parentNode === document.body && showMask) {
      const eventArray = ['touchstart'];
      const domArray = [
        document.body,
        this.maskDom,
        this.handlerDom,
        this.contentDom,
      ];
      if (open && document.body.style.overflow !== 'hidden') {
        if (right) {
          this.addScrollingEffect(right);
        }
        if (openCount === 1) {
          document.body.style.overflow = 'hidden';
        }
        document.body.style.touchAction = 'none';
        domArray.forEach((item, i) => {
          if (!item) {
            return;
          }
          addEventListener(
            item,
            eventArray[i] || 'touchmove',
            i ? this.removeMoveHandler : this.removeStartHandler,
            this.passive,
          );
        });
      } else if (this.getCurrentDrawerSome()) {
        // overflow;
        if (!openCount) {
          document.body.style.overflow = '';
        }
        document.body.style.touchAction = '';
        if (right) {
          this.remScrollingEffect(right);
        }
        domArray.forEach((item, i) => {
          if (!item) {
            return;
          }
          removeEventListener(
            item,
            eventArray[i] || 'touchmove',
            i ? this.removeMoveHandler : this.removeStartHandler,
            this.passive,
          );
        });
      }
    }
  };

  private addScrollingEffect = (right: number) => {
    const {
      placement,
      duration,
      ease,
      getOpenCount,
      switchScrollingEffect,
    } = this.props;
    const openCount = getOpenCount && getOpenCount();
    if (openCount === 1) {
      switchScrollingEffect();
    }
    const widthTransition = `width ${duration} ${ease}`;
    const transformTransition = `transform ${duration} ${ease}`;
    this.dom.style.transition = 'none';
    switch (placement) {
      case 'right':
        this.dom.style.transform = `translateX(-${right}px)`;
        break;
      case 'top':
      case 'bottom':
        this.dom.style.width = `calc(100% - ${right}px)`;
        this.dom.style.transform = 'translateZ(0)';
        break;
      default:
        break;
    }
    clearTimeout(this.timeout);
    this.timeout = setTimeout(() => {
      if (this.dom) {
        this.dom.style.transition = `${transformTransition},${widthTransition}`;
        this.dom.style.width = '';
        this.dom.style.transform = '';
      }
    });
  };

  private remScrollingEffect = (right: number) => {
    const {
      placement,
      duration,
      ease,
      getOpenCount,
      switchScrollingEffect,
    } = this.props;
    const openCount = getOpenCount && getOpenCount();
    if (!openCount) {
      switchScrollingEffect(true);
    }
    if (transitionStr) {
      document.body.style.overflowX = 'hidden';
    }
    this.dom.style.transition = 'none';
    let heightTransition: string;
    let widthTransition = `width ${duration} ${ease}`;
    const transformTransition = `transform ${duration} ${ease}`;
    switch (placement) {
      case 'left': {
        this.dom.style.width = '100%';
        widthTransition = `width 0s ${ease} ${duration}`;
        break;
      }
      case 'right': {
        this.dom.style.transform = `translateX(${right}px)`;
        this.dom.style.width = '100%';
        widthTransition = `width 0s ${ease} ${duration}`;
        if (this.maskDom) {
          this.maskDom.style.left = `-${right}px`;
          this.maskDom.style.width = `calc(100% + ${right}px)`;
        }
        break;
      }
      case 'top':
      case 'bottom': {
        this.dom.style.width = `calc(100% + ${right}px)`;
        this.dom.style.height = '100%';
        this.dom.style.transform = 'translateZ(0)';
        heightTransition = `height 0s ${ease} ${duration}`;
        break;
      }
      default:
        break;
    }
    clearTimeout(this.timeout);
    this.timeout = setTimeout(() => {
      if (this.dom) {
        this.dom.style.transition = `${transformTransition},${
          heightTransition ? `${heightTransition},` : ''
        }${widthTransition}`;
        this.dom.style.transform = '';
        this.dom.style.width = '';
        this.dom.style.height = '';
      }
    });
  };

  private getCurrentDrawerSome = () =>
    !Object.keys(currentDrawer).some(key => currentDrawer[key]);

  private getLevelDom = ({ level, getContainer }: IDrawerChildProps) => {
    if (windowIsUndefined) {
      return;
    }
    const container = getContainer && getContainer();
    const parent = container ? (container.parentNode as HTMLElement) : null;
    this.levelDom = [];
    if (level === 'all') {
      const children: HTMLElement[] = parent
        ? Array.prototype.slice.call(parent.children)
        : [];
      children.forEach((child: HTMLElement) => {
        if (
          child.nodeName !== 'SCRIPT' &&
          child.nodeName !== 'STYLE' &&
          child.nodeName !== 'LINK' &&
          child !== container
        ) {
          this.levelDom.push(child);
        }
      });
    } else if (level) {
      dataToArray(level).forEach(key => {
        document.querySelectorAll(key).forEach(item => {
          this.levelDom.push(item);
        });
      });
    }
  };

  private getHorizontalBoolAndPlacementName = () => {
    const { placement } = this.props;
    const isHorizontal = placement === 'left' || placement === 'right';
    const placementName = `translate${isHorizontal ? 'X' : 'Y'}`;
    return {
      isHorizontal,
      placementName,
    };
  };

  // tslint:disable-next-line:member-ordering
  public render() {
    const {
      className,
      children,
      style,
      width,
      height,
      defaultOpen,
      open: $open,
      prefixCls,
      placement,
      level,
      levelMove,
      ease,
      duration,
      getContainer,
      handler,
      onChange,
      afterVisibleChange,
      showMask,
      maskClosable,
      maskStyle,
      onClose,
      onHandleClick,
      keyboard,
      getOpenCount,
      switchScrollingEffect,
      ...props
    } = this.props;
    const open = this.dom ? $open : false;
    const wrapperClassName = classnames(prefixCls, {
      [`${prefixCls}-${placement}`]: true,
      [`${prefixCls}-open`]: open,
      [className || '']: !!className,
      'no-mask': !showMask,
    });

    const { placementName } = this.getHorizontalBoolAndPlacementName();
    // const defaultValue = !this.contentDom || !level ? '100%' : `${value}px`;
    const placementPos =
      placement === 'left' || placement === 'top' ? '-100%' : '100%';
    const transform = open ? '' : `${placementName}(${placementPos})`;

    const handlerChildren =
      handler &&
      React.cloneElement(handler, {
        onClick: (e: React.MouseEvent) => {
          if (handler.props.onClick) {
            handler.props.onClick();
          }
          if (onHandleClick) {
            onHandleClick(e);
          }
        },
        ref: (c: HTMLElement) => {
          this.handlerDom = c;
        },
      });
    return (
      <div
        {...props}
        tabIndex={-1}
        className={wrapperClassName}
        style={style}
        ref={(c: HTMLElement | null) => {
          this.dom = c as HTMLElement;
        }}
        onKeyDown={open && keyboard ? this.onKeyDown : undefined}
        onTransitionEnd={this.onWrapperTransitionEnd}
      >
        {showMask && (
          <div
            className={`${prefixCls}-mask`}
            onClick={maskClosable ? onClose : undefined}
            style={maskStyle}
            ref={c => {
              this.maskDom = c as HTMLElement;
            }}
          />
        )}
        <div
          className={`${prefixCls}-content-wrapper`}
          style={{
            transform,
            msTransform: transform,
            width: isNumeric(width) ? `${width}px` : width,
            height: isNumeric(height) ? `${height}px` : height,
          }}
          ref={c => {
            this.contentWrapper = c as HTMLElement;
          }}
        >
          <div
            className={`${prefixCls}-content`}
            ref={c => {
              this.contentDom = c as HTMLElement;
            }}
            onTouchStart={
              open && showMask ? this.removeStartHandler : undefined
            } // 
            onTouchMove={open && showMask ? this.removeMoveHandler : undefined} //
          >
            {children}
          </div>
          {handlerChildren}
        </div>
      </div>
    );
  }
}

export default polyfill(DrawerChild);
