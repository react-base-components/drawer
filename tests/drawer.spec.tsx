// eslint-disable react/no-multi-comp
import { mount } from 'enzyme';
import * as React from 'react';
import toJson from 'enzyme-to-json';
import Drawer from "../src";
import { IDrawerProps } from '../src/IDrawerPropTypes';


class DrawerTesterRef extends React.Component {
  public container: HTMLDivElement;

  public getContainer = () => this.container;

  public saveContainer = (container: HTMLDivElement) => {
    this.container = container;
  };

  public render() {
    return (
      <div>
        <div ref={this.saveContainer} className="main" />
        <Drawer {...this.props} open getContainer={this.getContainer}>
          <p className="text">Here is content of Drawer</p>
        </Drawer>
      </div>
    );
  }
}
/* eslint react/no-multi-comp: 0 */

interface IState {
  visible: boolean;
}
// tslint:disable-next-line:max-classes-per-file
class DrawerTesterDom extends React.Component<IDrawerProps, IState> {
  public container: HTMLDivElement;

  constructor(props: IDrawerProps) {
    super(props);
    this.state = { visible: false };
  }

  public componentDidMount() {
    this.setState({ visible: true }); // eslint-disable-line react/no-did-mount-set-state
  }

  public getContainer = () => this.container;

  public saveContainer = (container: HTMLDivElement) => {
    this.container = container;
  };

  public render() {
    return (
      <div>
        <div ref={this.saveContainer} className="main" />
        {this.state.visible ? (
          <Drawer {...this.props} open getContainer={this.getContainer()}>
            <p className="text">Here is content of Drawer</p>
          </Drawer>
        ) : null}
      </div>
    );
  }
}

/* eslint react/no-multi-comp: 0 */
// tslint:disable-next-line:max-classes-per-file
const DrawerTesterBoolean = props => (
  <div>
    <Drawer {...props} open getContainer={false}>
      <p className="text">Here is content of Drawer</p>
    </Drawer>
  </div>
);

describe('Drawer', () => {
  it('render function', () => {
    const wrapper = mount(<DrawerTesterRef />);
    expect(toJson(wrapper.render())).toMatchSnapshot();
  });

  it('render dom', () => {
    const wrapper = mount(<DrawerTesterDom />);
    expect(toJson(wrapper.render())).toMatchSnapshot();
  });

  it('render boolean', () => {
    const wrapper = mount(<DrawerTesterBoolean />);
    expect(toJson(wrapper.render())).toMatchSnapshot();
  });
});
