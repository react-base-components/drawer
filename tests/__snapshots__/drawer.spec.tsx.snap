// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer render boolean 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="drawer drawer-left drawer-open"
      tabindex="-1"
    >
      <div
        class="drawer-mask"
      />
      <div
        class="drawer-content-wrapper"
        style=""
      >
        <div
          class="drawer-content"
        >
          <p
            class="text"
          >
            Here is content of Drawer
          </p>
        </div>
        <div
          class="drawer-handle"
        >
          <i
            class="drawer-handle-icon"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Drawer render dom 1`] = `
<div>
  <div
    class="main"
  >
    <div>
      <div
        class="drawer drawer-left drawer-open"
        tabindex="-1"
      >
        <div
          class="drawer-mask"
        />
        <div
          class="drawer-content-wrapper"
          style=""
        >
          <div
            class="drawer-content"
          >
            <p
              class="text"
            >
              Here is content of Drawer
            </p>
          </div>
          <div
            class="drawer-handle"
          >
            <i
              class="drawer-handle-icon"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Drawer render function 1`] = `
<div>
  <div
    class="main"
  >
    <div>
      <div
        class="drawer drawer-left drawer-open"
        tabindex="-1"
      >
        <div
          class="drawer-mask"
        />
        <div
          class="drawer-content-wrapper"
          style=""
        >
          <div
            class="drawer-content"
          >
            <p
              class="text"
            >
              Here is content of Drawer
            </p>
          </div>
          <div
            class="drawer-handle"
          >
            <i
              class="drawer-handle-icon"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
